File: 3.Ottobre/60.SpaceWeekDay6-MoonPhase-09Ottobre.js
//todo  Per il sesto giorno della Settimana Spaziale, ti verrà fornita una data nel formato "AAAA-MM-GG" e dovrai determinare la fase lunare per quel giorno utilizzando le seguenti regole:

//! Utilizza un ciclo lunare semplificato di 28 giorni, suddiviso in quattro fasi uguali:

//! "Nuova": giorni 1 - 7
//! "Crescente": giorni 8 - 14
//! "Piena": giorni 15 - 21
//! "Calante": giorni 22 - 28
//! Dopo il giorno 28, il ciclo si ripete con il giorno 1, una luna nuova.

//! Utilizza "2000-01-06" come luna nuova di riferimento (giorno 1 del ciclo) per determinare la fase del giorno specificato.
//! Non ti verranno fornite date precedenti alla data di riferimento.
//! Restituisci la fase corretta come stringa.
//! Nota: il giorno 1 rappresenta il giorno della luna nuova, il che significa che sono trascorsi 0 giorni dall'ultima luna nuova.

//? Tempo di esecuzione:
//* Non completato

//?Codice scritto senza utilizzo di ChatGpt

function moonPhase(dateString) {

  let day = dateString.split("").splice(8).join("")

  if(/00-07/.test(day)) return "New"
  if(/08-14/.test(day)) return "Waxing"
  if(/15-21/.test(day)) return "Full"
  if(/22-28/.test(day)) return "Waning"
  
}

console.log(moonPhase("2000-01-12"))

//?Codice scritto da ChatGpt

function moonPhase(dateString) {
  const reference = new Date("2000-01-06");
  const date = new Date(dateString);

  const daysPassed = Math.floor(
    (date - reference) / (1000 * 60 * 60 * 24)
  );

  const cycleDay = (daysPassed % 28) + 1;

  if (cycleDay <= 7) return "New";
  if (cycleDay <= 14) return "Waxing";
  if (cycleDay <= 21) return "Full";
  return "Waning";
}

console.log(moonPhase("2000-01-12"));


//* ✅ Pro

//* Logica corretta rispetto al problema

//* Usa la data di riferimento giusta

//* Gestisce il ciclo lunare che si ripete

//* Codice leggibile e facile da seguire

//* Nessun trucco con regex inutili

//! ❌ Contro

//! Dipende da Date (attenzione ai fusi orari)

//! Math.floor può dare differenze se l’orario non è UTC

//! Un po’ più lungo rispetto a una soluzione “naive”

//? Versione Chat Gpt 

//todo Considerazioni: 

//* Qui ho proprio sbagliato la consegna

//! N.B. LEGGI MEGLIO LA CONSEGNA E IMPARA IL REGEX

File: Python/3.Ottobre/60.Space_week_day6-Moon_phase-09Ottobre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# //todo Consegna dell'esercizio

# Per il sesto giorno della Space Week, ti verrà fornita una data
# nel formato "YYYY-MM-DD" e dovrai determinare la fase lunare.
#
# Utilizza un ciclo lunare semplificato di 28 giorni:
#
# "New"    : giorni 1 - 7
# "Waxing" : giorni 8 - 14
# "Full"   : giorni 15 - 21
# "Waning" : giorni 22 - 28
#
# Usa "2000-01-06" come riferimento per la luna nuova (giorno 1).
# Non verranno fornite date precedenti alla data di riferimento.
#
# Restituisci la fase corretta come stringa.

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

import re

def moon_phase(date_string):
    day = int(date_string[8:])

    if day <= 7:
         return "New"
    if day <= 14:
         return "Waxing"
    if day <= 21:
         return "Full"
    if day <= 28:
         return "Waning"

print(moon_phase("2000-01-12"))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

from datetime import datetime

def moon_phase(date_string):
    reference_date = datetime.strptime("2000-01-06", "%Y-%m-%d")
    current_date = datetime.strptime(date_string, "%Y-%m-%d")

    days_passed = (current_date - reference_date).days
    cycle_day = (days_passed % 28) + 1

    if cycle_day <= 7:
        return "New"
    elif cycle_day <= 14:
        return "Waxing"
    elif cycle_day <= 21:
        return "Full"
    else:
        return "Waning"

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Hai individuato correttamente le quattro fasce di giorni associate
#   alle fasi lunari.
# - La logica degli if è semplice e leggibile.
# - Hai estratto correttamente il giorno dalla stringa usando lo slicing.

# //! Contro
# - La consegna richiede di calcolare i giorni trascorsi dalla data
#   di riferimento, non di usare il giorno del mese.
# - Il ciclo lunare di 28 giorni non viene considerato.
# - import re è inutilizzato.
# - La variabile day_join non esiste e genera un errore.
# - Funziona solo per alcune date dello stesso mese, ma fallisce
#   appena cambia mese o anno.

# =========================================================
# //* Consigli
# =========================================================

# - Quando una consegna parla di date, valuta subito l'utilizzo del
#   modulo datetime.
# - Leggi attentamente eventuali date di riferimento fornite dal testo:
#   spesso sono la chiave dell'esercizio.
# - Elimina gli import inutilizzati.
# - Prima di consegnare il codice, verifica che tutte le variabili
#   utilizzate siano state definite.
# - Testa sempre il programma con date di mesi e anni diversi per
#   verificare che la logica sia davvero generale.